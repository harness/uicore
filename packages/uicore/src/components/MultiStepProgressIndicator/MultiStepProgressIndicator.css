/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

:root {
  --barWidth: 25px;
}

.dotOutLine {
  outline: 1px solid var(--grey-200);
}

.dot {
  width: 20px;
  height: 20px;
  border-radius: 10px;
  border: 3px solid var(--white);
  background: var(--grey-300);
  box-sizing: border-box;
}

.dotSuccess {
  background: var(--primary-6);
  border: 3px solid var(--primary-4);
  outline: 0.1px solid var(--primary-4);
}

.halfBar {
  width: var(--barWidth);
  border-top-right-radius: 5px;
  border-bottom-right-radius: 5px;
}

.fullBar {
  width: calc(2 * (var(--barWidth)));
}

.statusBar {
  height: 5px;
  width: calc(2 * (var(--barWidth)));
  position: relative;
}

.bar {
  height: 5px;
  position: absolute;
  background: var(--grey-200);
}

.barSuccess {
  height: 5px;
  background: linear-gradient(var(--teal-300), var(--primary-6));
}

.spacing {
  padding: 2px !important;
}
