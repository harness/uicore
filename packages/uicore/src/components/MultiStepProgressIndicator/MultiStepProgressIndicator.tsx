/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import React from 'react'
import cx from 'classnames'

import { Layout } from '../../layouts/Layout'
import { Text } from '../..'

import css from './MultiStepProgressIndicator.css'
import { Color } from '@harness/design-system'

type StepStatus = 'TODO' | 'INPROGRESS' | 'FAILED' | 'SUCCESS'

export interface MultiStepProgressIndicatorProps {
  progressMap: Map<number, { StepStatus: StepStatus; StepName?: string }>
}

const Dot: React.FC<{ status: StepStatus; name?: string }> = ({ status, name }) => {
  switch (status) {
    case 'TODO':
    case 'FAILED':
      return (
        <Layout.Vertical flex>
          <div className={css.dotNameFailed}>
            <Text lineClamp={1} className={css.textmaxlength}>
              {name}
            </Text>
          </div>
          <div className={cx(css.dot, css.dotOutLine, css.spacing)} />
        </Layout.Vertical>
      )
    case 'INPROGRESS':
    case 'SUCCESS':
      return (
        <Layout.Vertical flex>
          <div className={css.dotNameSuccess}>
            <Text lineClamp={1} className={css.textmaxlength} color={Color.PRIMARY_7}>
              {name}
            </Text>
          </div>
          <div className={cx(css.dot, css.dotSuccess, css.spacing)} />
        </Layout.Vertical>
      )
    default:
      return null
  }
}

const Bar: React.FC<{ status: StepStatus }> = ({ status }) => {
  switch (status) {
    case 'TODO':
    case 'FAILED':
      return (
        <div className={css.statusBar}>
          <div className={cx(css.bar, css.fullBar)} />
        </div>
      )
    case 'INPROGRESS':
      return (
        <div className={css.statusBar}>
          <div className={cx(css.bar, css.fullBar)} />
          <div className={cx(css.bar, css.barSuccess, css.halfBar)} />
        </div>
      )
    case 'SUCCESS':
      return (
        <div className={css.statusBar}>
          <div className={cx(css.bar, css.barSuccess, css.fullBar)} />
        </div>
      )
    default:
      return null
  }
}

export const MultiStepProgressIndicator: React.FC<MultiStepProgressIndicatorProps> = ({ progressMap }) => {
  const entries = progressMap.size ? Array.from(progressMap.entries()) : []

  const elements: React.ReactNode[] = []

  entries.forEach((value, index) => {
    const status = value[1].StepStatus
    const name = value[1].StepName

    elements.push(
      <Layout.Horizontal flex key={index}>
        <Dot status={status} name={name} />
        {index !== entries.length - 1 ? <Bar status={status} /> : null}
      </Layout.Horizontal>
    )
  })

  return <Layout.Horizontal>{elements}</Layout.Horizontal>
}
