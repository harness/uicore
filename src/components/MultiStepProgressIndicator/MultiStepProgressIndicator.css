/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

.dot {
  width: 20px;
  height: 20px;
  border-radius: 10px;
  background: var(--grey-200);
  border: 1.35714px solid var(--white);
  box-sizing: border-box;
}

.bar {
  width: 23px;
  height: 3px;
  background: var(--grey-200);
  border-radius: 10px;
}

.barSuccess {
  background: var(--green-600);
}

.spacing {
  padding: 2px !important;
}
