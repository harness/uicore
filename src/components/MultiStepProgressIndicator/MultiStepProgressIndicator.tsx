/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import React from 'react'
import cx from 'classnames'

import { Layout } from '../../layouts/Layout'
import { Icon } from '@harness/icons'
import { Color } from '@harness/design-system'

import css from './MultiStepProgressIndicator.css'

type StepStatus = 'TODO' | 'INPROGRESS' | 'FAILED' | 'SUCCESS'

export interface MultiStepProgressIndicatorProps {
  progressMap: Map<number, StepStatus>
}

const Dot: React.FC<{ status: StepStatus }> = ({ status }) => {
  switch (status) {
    case 'TODO':
      return <div className={cx(css.dot, css.spacing)} />
    case 'INPROGRESS':
      return <Icon name="steps-spinner" size={20} color={Color.GREEN_600} className={css.spacing} />
    case 'FAILED':
      return <Icon name="circle-cross" size={20} color={Color.RED_600} className={css.spacing} />
    case 'SUCCESS':
      return <Icon name="success-tick" size={22} />
    default:
      return null
  }
}

const Bar: React.FC<{ status: StepStatus }> = ({ status }) => {
  switch (status) {
    case 'TODO':
    case 'INPROGRESS':
    case 'FAILED':
      return <div className={css.bar} />
    case 'SUCCESS':
      return <div className={cx(css.bar, css.barSuccess)} />
    default:
      return null
  }
}

export const MultiStepProgressIndicator: React.FC<MultiStepProgressIndicatorProps> = ({ progressMap }) => {
  const entries = progressMap.size ? Array.from(progressMap.entries()) : []
  const elements: React.ReactNode[] = []

  entries.forEach((value, index) => {
    const status = value[1]
    elements.push(
      <Layout.Horizontal flex key={index}>
        <Dot status={status} />
        {index !== progressMap.size - 1 ? <Bar status={status} /> : null}
      </Layout.Horizontal>
    )
  })

  return <Layout.Horizontal>{elements}</Layout.Horizontal>
}
