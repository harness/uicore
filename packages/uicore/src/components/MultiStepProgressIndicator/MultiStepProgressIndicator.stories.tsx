/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import React from 'react'
import type { Meta, Story } from '@storybook/react'
import { Title, Primary } from '@storybook/addon-docs/blocks'
import { Text } from '../..'

import { MultiStepProgressIndicator, MultiStepProgressIndicatorProps } from './MultiStepProgressIndicator'

export default {
  title: 'Components / MultiStepProgressIndicator',
  component: MultiStepProgressIndicator,
  parameters: {
    layout: 'centered',
    docs: {
      source: {
        type: 'code'
      },

      page: function PageDescription() {
        return (
          <>
            <Title>Multi-Step Progress Indicator</Title>
            This component renders sequential step progress.
            <Primary />
            <Text style={{ fontWeight: 'bold' }}>Data Interface</Text>
            <pre>
              <code>{`type StepStatus = 'TODO' | 'INPROGRESS' | 'FAILED' | 'SUCCESS'`}</code>
              <br />
              <br />
              <code>
                {`interface MultiStepProgressIndicatorProps {
  progressMap: Map<number, {StepStatus:'',StepName:''}>
  
}
`}
              </code>
            </pre>
          </>
        )
      }
    }
  }
} as Meta

export const MultiStepProgressIndicatorComp: Story<MultiStepProgressIndicatorProps> = args => (
  <MultiStepProgressIndicator {...args} />
)

MultiStepProgressIndicatorComp.args = {
  progressMap: new Map([
    [0, { StepStatus: 'SUCCESS', StepName: 'Workload' }],
    [1, { StepStatus: 'SUCCESS', StepName: 'Artifact' }],
    [2, { StepStatus: 'INPROGRESS', StepName: 'Infrastructure' }],
    [3, { StepStatus: 'TODO', StepName: 'Pipeline' }]
  ])
}
