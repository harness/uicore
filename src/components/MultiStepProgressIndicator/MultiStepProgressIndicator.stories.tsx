/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import React from 'react'
import type { Meta, Story } from '@storybook/react'
import { Title, Primary } from '@storybook/addon-docs/blocks'
import { Text } from '../..'

import { MultiStepProgressIndicator, MultiStepProgressIndicatorProps } from './MultiStepProgressIndicator'

export default {
  title: 'Components / MultiStepProgressIndicator',
  component: MultiStepProgressIndicator,
  parameters: {
    layout: 'centered',
    docs: {
      source: {
        type: 'code'
      },

      page: function PageDescription() {
        return (
          <>
            <Title>Multi-Step Progress Indicator</Title>
            This component renders sequential step progress.
            <Primary />
            <Text style={{ fontWeight: 'bold' }}>Data Interface</Text>
            <pre>
              <code>
                {`interface StepDetails {
  stepIndex: number
  stepStatus: 'TODO' | 'INPROGRESS' | 'FAILED' | 'SUCCESS'
}`}
              </code>
              <br />
              <br />
              <code>
                {`interface MultiStepProgressIndicatorProps {
   stepProgress: StepDetails[]
}`}
              </code>
            </pre>
          </>
        )
      }
    }
  }
} as Meta

export const MultiStepProgressIndicatorComp: Story<MultiStepProgressIndicatorProps> = args => (
  <MultiStepProgressIndicator {...args} />
)

MultiStepProgressIndicatorComp.args = {
  stepProgress: [
    { stepIndex: 0, stepStatus: 'SUCCESS' },
    { stepIndex: 1, stepStatus: 'FAILED' },
    { stepIndex: 2, stepStatus: 'INPROGRESS' },
    { stepIndex: 3, stepStatus: 'TODO' }
  ]
}
