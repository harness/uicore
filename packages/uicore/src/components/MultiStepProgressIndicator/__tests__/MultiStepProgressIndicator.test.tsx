/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import React from 'react'
import { render } from '@testing-library/react'
import { MultiStepProgressIndicator } from '../MultiStepProgressIndicator'

describe('Test MultiStepProgressIndicator', () => {
  test('Basic render', () => {
    const { container } = render(
      <MultiStepProgressIndicator
        progressMap={
          new Map([
            [0, { StepStatus: 'SUCCESS', StepName: 'StepName1' }],
            [1, { StepStatus: 'SUCCESS', StepName: 'StepName2' }],
            [2, { StepStatus: 'INPROGRESS', StepName: 'StepName3' }],
            [3, { StepStatus: 'TODO', StepName: 'StepName4' }]
          ])
        }
      />
    )

    // show stepname
    expect(document.body.getElementsByClassName('dotNameSuccess').length).toBe(3)
    expect(document.body.getElementsByClassName('dotNameFailed').length).toBe(1)

    //  should  show 1 grey dot
    expect(document.body.getElementsByClassName('dot dotOutLine spacing').length).toBe(1)
    // should show 3 blue success dot
    expect(document.body.getElementsByClassName('dot dotSuccess spacing').length).toBe(3)

    // show should  full bar which will be overlapped by half bar if successfull
    expect(document.body.getElementsByClassName('bar fullBar').length).toBe(3)

    // should show 1 blue success half bar
    expect(document.body.getElementsByClassName('bar barSuccess halfBar').length).toBe(1)

    expect(container).toMatchSnapshot()
  })
})
