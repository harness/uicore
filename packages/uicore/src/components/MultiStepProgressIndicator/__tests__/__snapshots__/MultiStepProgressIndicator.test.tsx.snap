// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test MultiStepProgressIndicator Basic render 1`] = `
<div>
  <div
    class="horizontal layout-spacing-undefined StyledProps--main"
  >
    <div
      class="horizontal layout-spacing-undefined StyledProps--main StyledProps--flex"
    >
      <div
        class="dot dotSuccess spacing"
      />
    </div>
    <div
      class="horizontal layout-spacing-undefined StyledProps--main StyledProps--flex"
    >
      <div
        class="statusBar"
      >
        <div
          class="bar fullBar"
        />
      </div>
      <div
        class="dot dotOutLine spacing"
      />
    </div>
    <div
      class="horizontal layout-spacing-undefined StyledProps--main StyledProps--flex"
    >
      <div
        class="statusBar"
      >
        <div
          class="bar fullBar"
        />
        <div
          class="bar barSuccess halfBar"
        />
      </div>
      <div
        class="dot dotOutLine spacing"
      />
    </div>
    <div
      class="horizontal layout-spacing-undefined StyledProps--main StyledProps--flex"
    >
      <div
        class="statusBar"
      >
        <div
          class="bar fullBar"
        />
      </div>
      <div
        class="dot dotOutLine spacing"
      />
    </div>
  </div>
</div>
`;
